# Programmation - Python 3

## Les boucles

<span style='font-size:22px; '>&#128203;</span>
Retenez ceci :

Une boucle permet de **répéter** un certain nombre de fois une ou plusieurs instructions.

Il y deux types de boucles:
* les boucles bornées
* les boucles non bornées

Dans **une boucle bornée**, on connaît à l'avance le nombre de répétitions à faire.<br>
On les écrits en langage courant de la manière suivante:

**Pour** i allant de $0$ à $n$ par pas de $1$ **faire** ceci ou cela.<br>
Cette boucle se répète $n+1$ fois, car de 0 à $n$,  il y a $n+1$ entiers (le pas ici est de $1$).

Dans **une boucle non bornée**, on ne connaît pas à l'avance le nombre de répétitions à faire.<br>
Celles-ci dépendent d'une condition.<br>
En langage courant cela s'écrit:

**Tant que** une ou plusieurs conditions sont vraies(ou fausses) **faire** ceci ou cela.<br>
Le nombre de répétitions dépend de conditions.


## Les boucles bornées en Python

**La syntaxe :**

```py
for i in range(debut,fin,pas):
    instructions
```
Cette boucle fait varier i entre debut et fin sans atteindre fin...<br>
1er tour : i = debut<br>
2ème tour : i = debut + pas<br>
3ème tour : i = debut + pas + pas<br>
etc..<br>
**Si l'ajout d'un pas est supérieur ou égal à fin, la boucle se termine.**


***Exemples***
```py
for i in range(0,10,1):
    instructions
```
i vaudra 0,1,2,3,4,5,6,7,8 et 9 (en tout 10 répétitions)

```py
for i in range(0,10,2):
    instructions
```
i vaudra 0,2,4,6 et 8 (en tout 5 répétitions)

```py
for i in range(-5,10,3):
    instructions
```
i vaudra -5,-2,1,4 et 7 (en tout 5 répétitions)

```py
for i in range(10,-1,-2):
    instructions
```
i vaudra 10,8,6,4,2 et 0 (en tout 6 répétitions)

Quand on souhaite faire n répétitions, il suffit d'écrire:
```py
for i in range(n):
    instructions
```
En Python range(n) équivaut à range(0,n,1).

Dans l'exemple ci-dessous, on fait afficher une table de multiplication:

In [3]:
for i in range(12):
    print(i," x 7 = ",7*i)

0  x 7 =  0
1  x 7 =  7
2  x 7 =  14
3  x 7 =  21
4  x 7 =  28
5  x 7 =  35
6  x 7 =  42
7  x 7 =  49
8  x 7 =  56
9  x 7 =  63
10  x 7 =  70
11  x 7 =  77


### Exercice 1


Faire afficher une autre table de multiplication

In [ ]:
# code ici

### Exercice 2

Faire afficher avec une boucle les 25 premiers nombres impairs<br>

Résultat attendu:
1 3 5 7 9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49 

In [4]:
# code ici
    

### Exercice 3

Dans le code ci-dessous on utilise une boucle pour faire afficher le résultat de $a\times b$ où $a$ et $b$ sont des entiers naturels<br>
$a\times b$ signifie $\underbrace{a+a+...+a}_{\text{b fois}}$

In [7]:
a =  123
b = 20
produit = 0
for i in range(b):
    produit = produit + a
print(produit)
print(123*20)# pour vérifier

2460
2460


Modifier le programme précédent pour qu'il calcule le produit de deux entiers relatifs.<br>
C'est à dire, qu'il tienne compte des signes éventuels des entiers $a$ et $b$.
* $123 \times (-20) = -2460$
* $-123 \times (-20) = 2460$
* $-123 \times 20 = -2460$
* $123 \times 20 = 2460$

In [ ]:
# code ici

### Exercice 4

Dans le programme ci-dessous, on utilise une boucle pour faire afficher dans une fenêtre des lignes verticales

Orientation de la fenêtre
<pre>
(o,o)------------>(300,0) abscisses
  |
  |
  |
  |
  |
  |
  |
  v ------------->(300,300)
(0,300)
(ordonnées)
</pre>

In [15]:
from p5 import*

def setup():
    createCanvas(300,300)
    background(255,160,122)# couleur du fond de la fenêtre
    strokeWeight(2) # épaisseur des traits
    for i in range(0,301,20):
        line(i,0,i,300) # dessine une ligne entre les points de coordonnées (i,0) et (i,300)

        
def draw():
    pass

run()


Réaliser les figures suivantes :

<img src = "https://www.nsi-ljm.fr/images/boucles_exos.png" >

Pour des cercles l'instruction est :<br>
circle(abscisse_du_centre,ordonnée_du_centre,diamètre)

In [ ]:
# figure 2
from p5 import*

def setup():
    createCanvas(300,300)
    background(255,160,122)
    # à compléter
        
def draw():
    pass

run()

In [ ]:
# figure 3
from p5 import*

def setup():
    createCanvas(300,300)
    background(255,160,122)
    # à compléter
        
def draw():
    pass

run()

In [ ]:
# figure 4
from p5 import*

def setup():
    createCanvas(300,300)
    background(255,160,122)
    # à compléter
        
def draw():
    pass

run()

In [ ]:
# figure 5
from p5 import*

def setup():
    createCanvas(300,300)
    background(255,160,122)
    # à compléter
        
def draw():
    pass

run()

In [ ]:
# figure 6
from p5 import*

def setup():
    createCanvas(300,300)
    background(255,160,122)
    # à compléter
        
def draw():
    pass

run()

In [ ]:
# figure 7
from p5 import*

def setup():
    createCanvas(300,300)
    background(255,160,122)
    # à compléter
        
def draw():
    pass

run()

### Exercice 5

Les couleurs se représente par 3 entiers compris entre 0 et 255 (rouge, vert, bleu)
* (0,0,0) donne du noir
* (255,0,0) donne du rouge
* (0,255,0) donne du vert
* (0,0,255) donne du bleu
* (255,255,255) donne du blanc
* ( 125,200,83) donne une couleur...

Le programme ci-dessous affiche dans une fenêtre de dimension $255 \times 255$ pixels un dégradé de rouge.

In [19]:
from p5 import*

def setup():
    createCanvas(255,255)
    for i in range(1,256):
        stroke(i,0,0)
        strokeWeight(2)
        line(i,0,i,255)

def draw():
    pass

run()

Réaliser un dégradé de vert horizontalement

In [ ]:
from p5 import*

def setup():
    createCanvas(255,255)
    # à compléter

def draw():
    pass

run()

### Exercice 6

Le programme ci-dessous affiche un rectangle de dimensions et de couleur aléatoire dans une fenêtre.<br>
La couleur est donnée avec une transparence de 60.

Faites en sorte d'en afficher 500

In [ ]:
from p5 import*

def setup():
    createCanvas(300,300)
    noStroke()
    fill(random(255),random(255),random(255),60)
    rect(random(300),random(300),random(100),random(100))

def draw():
    pass

run()

## Les boucles non bornées en Python

**Syntaxe**
```py

while condition:
    instructions
```

Il faut bien sûr veiller à ce que la boucle s'arrête...<br>
C'est à dire que la condition doit à un certain moment être contredite...

**Exemples**<br>
Ce programme ne s'arrêtera jamais , car jamais i ne sera supérieur ou égal à 10.
```py
i = 0
while i < 10:
    print(i)
```
Celui là s'arrêtera quand i sera égal à 10

```py
i = 0
while i < 10:
    print(i)
    i = 
    i + 1
```

Le programme ci-dessous calcule la somme des 10 premiers entiers<br>
$0+1 +2+3+4+5+6+7+8+9=45$

In [81]:
i=0
somme=0
while i<10:
    somme = somme + i
    i = i + 1
print(somme)

45


### Exercice 7

Écrire un programme avec une boucle while, qui calcule la somme des 101 premiers entiers naturels.<br>
$0 + 1 +2+...+100$<br>
Résultat attendu : 5050



In [ ]:
# code ici


### Exercice 8

Écrire un programme avec une boucle while, qui calcule le produit des 10 premiers entiers naturels non nuls.<br>
$ 1 \times 2\times 3...\times 10$<br>
Résultat attendu : 3628800

In [ ]:
# code ici

### Exercice 9

Écrire un programme avec une boucle while, qui calcule la somme des entiers naturels pairs jusqu'à 100.<br>
$2+4+6+...+98+100$<br>
Résultat attendu : 2550

In [ ]:
# code ici

### Exercice 10

Écrire un programme avec une boucle while, qui calcule la somme des entiers naturels impairs jusqu'à 100.<br>
$1+3+5+...+99$<br>
Résultat attendu : 2500

In [ ]:
# code ici